### DISCUSSION TITLE
Does JBrowse support user to load their own custom track from vcf file

### DISCUSSION THREAD

**Discussion by JeanQ2020** [GMOD/jbrowse-components#4064] [General]
We'd like to find a browser which can allow user to load their own variant file (vcf) stored as temporary session and delete the track when user exits the web. I couldn't find any document that mentions this capacity for JBrowse

**Comment by cmdcolin** [comment]
@JeanQ2020 Yes, we do allow the user to open their own VCF or other types of files. If the file is large, I recommend using a bgzip+tabix indexed VCF, but for smallish VCF (think megabytes instead of gigabytes), then you can just open a plaintext VCF with the add track button.

I created a video to demonstrate how this works below


This video is using our "demo" browser on jbrowse.org but would apply to an instance if you set it up on your website too

The video shows the following things

a) adding a custom vcf file as a track. this file exists on the users hard drive. 
b) then reloading the page. the web browser doesn't allow us to keep that file open after a page reload, but the track still stays alive in the session. so I actually demo manually re-loading it via the Settings menu, then re-performing the Open file for the track. i'm not sure you want that feature or care about it, but it's just how we have it currently.
c) I then navigate to a new session (click the Browse demo button again, this starts a new "session") and the local file track is not there anymore. that is sort of equivalent to the track being deleted when the user exits the web.

video:

https://github.com/GMOD/jbrowse-components/assets/6511937/d421aa02-7b1e-4484-8458-9f61840d0dca

**Comment by JeanQ2020** [comment]
@cmdcolin Thank you so much for your quick response and helpful video. This is definitely what we'd like to see. I have two follow up questions: 1) does unix installation with web version also provide the same functionality for user to load and view their own custom vcf track on their own session (temporarily)? 2) does JBrowse allow long url with params (e.g., chr, coordinates)? We'd like to have ability to link a variant DB tool to JBrowse such that user can click a variant coordinate on that tool to launch and view that variant on JBrowse.

**Reply by cmdcolin** [reply]
1) @JeanQ2020 yep this is available with the web version that is installed via our quickstart. the npm packages e.g. @jbrowse/react-linear-genome-view also allow it. we call the files the user opens "session tracks"
2) yes we call this the url query param API...let me know if that helps! here is an example https://jbrowse.org/code/jb2/main/?config=test_data/volvox/config.json&loc=ctgA:1-800&assembly=volvox&tracks=gff3tabix_genes,volvox_filtered_vcf,volvox_microarray,volvox_cram

docs for the url params https://jbrowse.org/jb2/docs/urlparams/

**Reply by cmdcolin** [reply]
not sure if it is something that you particularly want but session tracks can also be added via the URL, that is documented at that link too

**Reply by JeanQ2020** [reply]
@cmdcolin great! another question: can user session be shared such that the user can access it from another tool with embedded JBrowse url + params?

**Reply by cmdcolin** [reply]
I think yes, we have a "Share" button where it can share the entire session with another user. 

![untitled(3)](https://github.com/GMOD/jbrowse-components/assets/6511937/f23d6365-9b44-4843-a0a3-fef59e2ea0f1)


unfortunately you can't just "copy what is in your url bar at any time and paste to another user" but the share button allows you to get a URL that you can share

**Reply by JeanQ2020** [reply]
@cmdcolin Thanks again for your helpful response! Do you have any recommendation which version of JBrowse to download if I don't have root privilege on unix server? Also JBrowse has been installed under /var/www/html as light weight plugin for another web tool, Tripal. I'm planning to install any version that you recommended at the same location with different name if you think it'll work.

**Reply by cmdcolin** [reply]
> Do you have any recommendation which version of JBrowse to download if I don't have root privilege on unix server? 

fundamentally, "jbrowse-web" is just a set of html, js, and css files that can be hosted by a static website. there is no server side code like cgi-bin, or php, that runs on the server side. so basically, if you can find a way to host some static files, that is a way to do it. 

 we generally recommend just putting jbrowse in the /var/www/html/ folder as it is 'familiar to many web server admins' :)

i recommend run e.g. 


```bash
# this gets the latest version, basically what i recommend
jbrowse create yourfolder

# move the folder to the web directory, feel free to rename the folder whatever you want
mv yourfolder /var/www/html/yourfolder


# then visit http://yourhost/yourfolder in the browser
# again replacing the yourfolder with whatever you want the folder to be named
```

then run the commands like 

```
jbrowse add-track yourfile.gff --out /var/www/html/yourfolder
```

you just have to figure out what works best regarding folder permissions. a lot of computers require sudo for accessing /var/www/html/ so if you want to chown a particular folder to you, or you just instead create that folder in your home directory and then pass it off to your admin to be moved to /var/www/html/ later on, that is fine.

**Reply by JeanQ2020** [reply]
@cmdcolin Thank you so much for very helpful information and suggestions. unfortunately our system doesn't support the command: jbrowse create yourfolder. got error:  jbrowse: command not found. Also can't run command: npm install. I don't have root privilege

**Reply by cmdcolin** [reply]
you can try prefixing any command in our docs with "npx" so type "npx @jbrowse/cli create yourfolder". The "npx" command is a "command runner" that will automatically download and run the jbrowse cli tool. I think it works without admin privileges (note that depending on your setup you may run into other challenges without admin permissions like setting up the http server but hopefully this helps)

**Reply by garrettjstevens** [reply]
I think the command would be `npx @jbrowse/cli create yourfolder`, since it needs the package name.

**Reply by cmdcolin** [reply]
thanks updated my comment too!

**Reply by JeanQ2020** [reply]
@cmdcolin Thanks for your helpful suggestions. Unfortunately it doesn't work for me probably due to permission issue.
<img width="506" alt="image001" src="https://github.com/GMOD/jbrowse-components/assets/61297590/53d315c6-b94a-42a4-bc5a-9df28ebced7a">

**Reply by JeanQ2020** [reply]
@cmdcolin BTW, we have a virtual wet server that hosts one of our genomics websites using Tripal. It has JBrowse plugin with the genome browse that gets information from its postgresql database. Are you familiar with the jbrowse plugin on Tripal and if you know that it supports loading local custom track? I didn't see the + sign any where on its browse. It'll be great if it can support loading custom tracks.

![image](https://github.com/GMOD/jbrowse-components/assets/61297590/4f5b720e-85ef-41dd-aef7-3b770ea56313)
